crates/ruma-state-res/tests/it/main.rs

```diff
@@ -0,0 +1,285 @@
+#[derive(Serialize)]
+struct ResolvedState {
+    kind: StateEventType,
+    state_key: String,
+    event_id: OwnedEventId,
+    content: JsonValue,
+}
+
+impl PartialEq for ResolvedState {
+    fn eq(&self, other: &Self) -> bool {
+        self.kind == other.kind
+            && self.state_key == other.state_key
+            && self.event_id == other.event_id
+    }
+}
+
+impl Eq for ResolvedState {}
+
+impl Ord for ResolvedState {
+    fn cmp(&self, other: &Self) -> Ordering {
+        Ordering::Equal
+            .then(self.kind.cmp(&other.kind))
+            .then(self.state_key.cmp(&other.state_key))
+            .then(self.event_id.cmp(&other.event_id))
+    }
+}
+
+impl PartialOrd for ResolvedState {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// Results of a test
+struct TestResults {
+    /// The resolved state of the room
+    resolved_state: BTreeSet<ResolvedState>,
+}
+
+/// Test a list of JSON files containing a list of PDUs and return the results
+///
+/// State resolution is run incrementally on each JSON file.
+fn run<P>(paths: &[P]) -> Result<TestResults, Box<dyn Error>>
+where
+    P: AsRef<Path>,
+{
+    // Run `cargo test -- --show-output` to view traces, set `RUST_LOG` to control filtering
+    let _subscriber = tracing_subscriber::fmt()
+        .with_env_filter(EnvFilter::from_default_env())
+        .with_test_writer()
+        .finish()
+        .set_default();
+
+    let (head, tail) = paths.split_first().ok_or("there should be at least one JSON file")?;
+
+    let path = PathBuf::from("tests/it/fixtures").join(head);
+
+    #[allow(clippy::disallowed_methods)]
+    let pdus = serde_json::from_reader::<_, Vec<Pdu>>(File::open(path)?)?;
+
+    let room_version = {
+        let create = pdus.first().ok_or("there should be at least one PDU")?.content();
+        serde_json::from_str::<ExtractRoomVersion>(create.get())
+            .map_err(|_| "the first PDU should be an m.room.create event")?
+            .room_version
+    };
+
+    let mut state_sets = Vec::new();
+    for pdu in &pdus {
+        let mut state_map = StateMap::new();
+        state_map.insert(
+            (
+                pdu.event_type().to_string().into(),
+                pdu.state_key().ok_or("all PDUs should be state events")?.to_owned(),
+            ),
+            pdu.event_id().clone(),
+        );
+        state_sets.push(state_map);
+    }
+
+    let mut pdus_by_id =
+        pdus.into_iter().map(|pdu| (pdu.event_id().to_owned(), pdu)).collect::<HashMap<_, _>>();
+
+    let mut auth_chain_sets = Vec::new();
+    for pdu in pdus_by_id.values() {
+        auth_chain_sets.push(auth_events_dfs(&pdus_by_id, pdu)?);
+    }
+
+    let fetch_event = |x: &EventId| pdus_by_id.get(x).cloned();
+
+    let mut resolved_state = resolve(&room_version, &state_sets, auth_chain_sets, fetch_event)?;
+
+    // Incrementally resolve subsequent files on top of the state resolved above
```

Comment:
Actually do we need to call `resolve` individually for each file? Shouldn't `resolve` work the same if we provide all the state maps at once as the `state_sets`?

Reply:
> Actually do we need to call `resolve` individually for each file?

Not necessarily, but it has the benefit of testing doing stateres incrementally rather than doing it all at once. Something I was considering was adding a way to tell `snapshot_test!()` whether to combine the PDUs in advance and resolve them at once or do them incrementally, so that it's possible to choose which way you want to test.

Reply:
That's a good idea. If we do that we could check that we get the same result in both cases.

Reply:
Good point actually: we could do this automatically for all tests transparently without creating extra snapshot files.

Reply:
Okay, we're doing that now.